misc/mempool.c

```diff
@@ -0,0 +1,346 @@
+ * they will fail, e.g. if there isn't a free block).  So that is the basic
+ * operation that needs synchronization, which we can do piecewise as needed in
+ * small one-block chunks to preserve latency.  At most (in free_block) a
+ * single locked operation consists of four bit sets and dlist removals. If the
+ * overall allocation operation fails, we just free the block we have (putting
+ * a block back into the list cannot fail) and return failure.
+ *
+ * For user mode compatible sys_mem_pool pools, a semaphore is used at the API
+ * level since using that does not introduce latency issues like locking
+ * interrupts does.
+ */
+
+static inline int pool_irq_lock(struct sys_mem_pool_base *p)
+{
+	if (p->flags & SYS_MEM_POOL_KERNEL) {
+		return irq_lock();
+	} else {
+		return 0;
+	}
+}
+
+static inline void pool_irq_unlock(struct sys_mem_pool_base *p, int key)
+{
+	if (p->flags & SYS_MEM_POOL_KERNEL) {
+		irq_unlock(key);
+	}
+}
+
+static void *block_alloc(struct sys_mem_pool_base *p, int l, size_t lsz)
+{
+	sys_dnode_t *block;
+	int key = pool_irq_lock(p);
+
+	block = sys_dlist_get(&p->levels[l].free_list);
+	if (block) {
+		clear_free_bit(p, l, block_num(p, block, lsz));
+	}
+	pool_irq_unlock(p, key);
+
+	return block;
+}
+
+static void block_free(struct sys_mem_pool_base *p, int level,
+			      size_t *lsizes, int bn)
+{
+	int i, key, lsz = lsizes[level];
+	void *block = block_ptr(p, lsz, bn);
+
+	key = pool_irq_lock(p);
+
+	set_free_bit(p, level, bn);
+
+	if (level && partner_bits(p, level, bn) == 0xf) {
+		for (i = 0; i < 4; i++) {
+			int b = (bn & ~3) + i;
+
+			clear_free_bit(p, level, b);
+			if (b != bn &&
+			    block_fits(p, block_ptr(p, lsz, b), lsz)) {
+				sys_dlist_remove(block_ptr(p, lsz, b));
+			}
+		}
+
+		pool_irq_unlock(p, key);
+
+		/* tail recursion! */
+		block_free(p, level-1, lsizes, bn / 4);
+		return;
+	}
+
+	if (block_fits(p, block, lsz)) {
+		sys_dlist_append(&p->levels[level].free_list, block);
+	}
+
+	pool_irq_unlock(p, key);
+}
+
+/* Takes a block of a given level, splits it into four blocks of the
+ * next smaller level, puts three into the free list as in
+ * block_free() but without the need to check adjacent bits or
+ * recombine, and returns the remaining smaller block.
+ */
+static void *block_break(struct sys_mem_pool_base *p, void *block, int l,
+				size_t *lsizes)
+{
+	int i, bn, key;
+
+	key = pool_irq_lock(p);
+
+	bn = block_num(p, block, lsizes[l]);
+
+	for (i = 1; i < 4; i++) {
+		int lbn = 4*bn + i;
+		int lsz = lsizes[l + 1];
+		void *block2 = (lsz * i) + (char *)block;
+
+		set_free_bit(p, l + 1, lbn);
+		if (block_fits(p, block2, lsz)) {
+			sys_dlist_append(&p->levels[l + 1].free_list, block2);
+		}
+	}
+
+	pool_irq_unlock(p, key);
+
+	return block;
+}
+
+int _sys_mem_pool_block_alloc(struct sys_mem_pool_base *p, size_t size,
+			      u32_t *level_p, u32_t *block_p, void **data_p)
+{
+	int i, from_l;
+	int alloc_l = -1, free_l = -1;
+	void *data;
+	size_t lsizes[p->n_levels];
```

Comment:
Do we have alloca support?

Reply:
this was copy paste from the k_mem_pool code.
good question. might be better to just set this to 16 which is the max value of n_levels